Implementing Remote Status API

gabrielboliveira opened the pull request:
Closes #163.

Tested on macOS 10.12.6, golang 1.8.3, building with OpenCL @ MacBook Pro.

Example of JSON output:

```json
{
    "validShares": 0,
    "staleShares": 0,
    "invalidShares": 0,
    "totalShares": 0,
    "sharesPerMinute": 0,
    "started": 1502252670,
    "uptime": 3,
    "devices": [
        {
            "index": 0,
            "deviceName": "Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz",
            "deviceType": "CPU",
            "hashRate": 0,
            "hashRateFormatted": "0H/s",
            "fanPercent": 0,
            "temperature": 0,
            "started": 1502252664
        },
        {
            "index": 1,
            "deviceName": "Iris Pro",
            "deviceType": "GPU",
            "hashRate": 0,
            "hashRateFormatted": "0H/s",
            "fanPercent": 0,
            "temperature": 0,
            "started": 1502252666
        },
        {
            "index": 2,
            "deviceName": "GeForce GT 750M",
            "deviceType": "GPU",
            "hashRate": 0,
            "hashRateFormatted": "0H/s",
            "fanPercent": 0,
            "temperature": 0,
            "started": 1502252668
        }
    ],
    "pool": {
        "started": 1502252670,
        "uptime": 3
    }
}
```

jcvernaleo commented on config.go:
```diff
@@ -59,6 +59,9 @@ type config struct {
 	MemProfile string `long:"memprofile" description:"Write mem profile to the specified file"`
 
+	// Monitor options
+	Monitor string `long:"monitor" description:"Enable a JSON API status monitor on given port -- NOTE port must be between 1024 and 65536"`
```
You don't need the note about port numbers since that is unrelated to this and is really just a general UNIX port thing (requiring additional privileges for ports <1024).

gabrielboliveira replied:
Do you think it's good to add a `short` config for this?

jolan replied:
I'd prefer if it were a string ```[]string```.  Maybe call it APIListeners and copy & paste the code from:

https://github.com/decred/dcrd/blob/539fb844cca27456ce02b5185c97e79e63183798/config.go#L748-L758

and replace RPCListeners with APIListeners.

We should give a bit more control in case someone is running on a box with multiple interfaces since there isn't any authentication.

jcvernaleo replied:
If the most obvious letter is still available, I'm always in favor of a short option.

jcvernaleo commented:
@jolan I made a few comments on this but overall it looks pretty good if you want to take a look at it in the near future too.

gabrielboliveira commented:
@jcvernaleo Thanks! I'll sure add a section about this on README.

jolan commented:
Unfortunately, I've been busy with higher priority projects and haven't had time to give it more than a quick look.

gabrielboliveira commented:
@jolan @jcvernaleo 

I've updated with the multi-address implementation. 

I've also updated the README to add the status API example, and also I improved a little bit the Building section, just a few topics fixes.

gabrielboliveira commented:
@jolan I agree with the warning message, but I'm not sure if it would be good to auto change the default port.

What you think on disabling auto-start? Monitor disabled by default.

jolan commented:
That sounds good to me.

gabrielboliveira commented:
@dajohi Not able to fix it during these weeks, working hard. Probably this weekend I'll be able to fix :)

gabrielboliveira commented:
@jolan @dajohi Fixed. Now it will warn if the monitor doesn't start up, and it will not autostart

jolan commented:
Thank you for your contribution.
